### PR TITLE
Port to qtpy to support qt5  

### PR DESCRIPTION
**Pull request by ericpre**
Continuation of #144 to port to qtpy to support qt5. This PR needs pyface and traitsui development version to work. It works fairly nicely with qt5 and qt4 -I haven't tried pyside.

### Progress:
- [x] port to qtpy
- [x] update to new-style signal-slot connection
- [x] workaround for the floating console with pyqt5 hidden after start
- [x] fix issue with progressbar
- [x] ~~fix issue with a plotting issue with qt5: when a signal is plotted, an additional call to `update()` is necessary, for example resizing the window is enough.~~ This is actually a ~~matplotlib~~ hyperspy issue with matpltolib versions 2.1 and 2.2 and it works fine with 2.0. This is now fixed in https://github.com/hyperspy/hyperspy/pull/1890.
- [x] ready for review

### CODE REVIEW

**Comment by ericpre** [comment]
The HyperSpy bundle (including the HyperSpyUI from this PR) build fine using the last winpython release. HyperSpyUI works fine, there is just one issue with the executable to start HyperSpyUI; using `python __main__.py` works fine.

**Comment by ericpre** [comment]
The issue mentioned above with plotting are not directly related to pyqt5 (at least not with pyqt5 in hyperspyUI) and I think this is going to be fixed in matplotlib or hyperspy (possibly with a temporary workaround).

**Comment by jat255** [comment]
@ericpre I just tested the new bundle edition, and it's not working for me. I'm getting the same issue as #155.

I started by opening the winpython command prompt and running `python -m hyperspyui`.

The following is output in the console, and it looks like `mrcz` is setting the `matplotlib` backend, which might be throwing things off:
``` 
C:\Users\jat\programs\HyperSpy WinPython Bundle v1.4-pyqt5\python-3.5.4.amd64\lib\site-packages\hyperspyui\mainwindowbase.py:26: UserWarning:
This call to matplotlib.use() has no effect because the backend has already
been chosen; matplotlib.use() must be called *before* pylab, matplotlib.pyplot,
or matplotlib.backends is imported for the first time.

The backend was *originally* set to 'Qt5Agg' by the following code:
  File "C:\Users\jat\programs\HyperSpy WinPython Bundle v1.4-pyqt5\python-3.5.4.amd64\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "C:\Users\jat\programs\HyperSpy WinPython Bundle v1.4-pyqt5\python-3.5.4.amd64\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\jat\programs\HyperSpy WinPython Bundle v1.4-pyqt5\python-3.5.4.amd64\lib\site-packages\hyperspyui\__main__.py", line 129, in <module>
    main()
  File "C:\Users\jat\programs\HyperSpy WinPython Bundle v1.4-pyqt5\python-3.5.4.amd64\lib\site-packages\hyperspyui\__main__.py", line 103, in main
    from hyperspyui.util import dummy_context_manager
  File "C:\Users\jat\programs\HyperSpy WinPython Bundle v1.4-pyqt5\python-3.5.4.amd64\lib\site-packages\hyperspyui\util.py", line 25, in <module>
    import hyperspy.components1d
  File "C:\Users\jat\programs\HyperSpy WinPython Bundle v1.4-pyqt5\python-3.5.4.amd64\lib\site-packages\hyperspy\components1d.py", line 34, in <module>
    from hyperspy._components.arctan import Arctan
  File "C:\Users\jat\programs\HyperSpy WinPython Bundle v1.4-pyqt5\python-3.5.4.amd64\lib\site-packages\hyperspy\_components\arctan.py", line 23, in <module>
    from hyperspy.component import Component
  File "C:\Users\jat\programs\HyperSpy WinPython Bundle v1.4-pyqt5\python-3.5.4.amd64\lib\site-packages\hyperspy\component.py", line 30, in <module>
    from hyperspy.defaults_parser import preferences
  File "C:\Users\jat\programs\HyperSpy WinPython Bundle v1.4-pyqt5\python-3.5.4.amd64\lib\site-packages\hyperspy\defaults_parser.py", line 29, in <module>
    from hyperspy.io_plugins import default_write_ext
  File "C:\Users\jat\programs\HyperSpy WinPython Bundle v1.4-pyqt5\python-3.5.4.amd64\lib\site-packages\hyperspy\io_plugins\__init__.py", line 22, in <module>
    from hyperspy.io_plugins import (msa, digital_micrograph, fei, mrc, mrcz, ripple,
  File "C:\Users\jat\programs\HyperSpy WinPython Bundle v1.4-pyqt5\python-3.5.4.amd64\lib\site-packages\hyperspy\io_plugins\mrcz.py", line 19, in <module>
    import mrcz as _mrcz
  File "C:\Users\jat\programs\HyperSpy WinPython Bundle v1.4-pyqt5\python-3.5.4.amd64\lib\site-packages\mrcz\__init__.py", line 7, in <module>
    try: from . import ReliablePy
  File "C:\Users\jat\programs\HyperSpy WinPython Bundle v1.4-pyqt5\python-3.5.4.amd64\lib\site-packages\mrcz\ReliablePy.py", line 47, in <module>
    import matplotlib.pyplot as plt
  File "C:\Users\jat\programs\HyperSpy WinPython Bundle v1.4-pyqt5\python-3.5.4.amd64\lib\site-packages\matplotlib\pyplot.py", line 72, in <module>
    from matplotlib.backends import pylab_setup
  File "C:\Users\jat\programs\HyperSpy WinPython Bundle v1.4-pyqt5\python-3.5.4.amd64\lib\site-packages\matplotlib\backends\__init__.py", line 14, in <module>
    line for line in traceback.format_stack()


  matplotlib.use('module://hyperspyui.mdi_mpl_backend')
```

When moving the little red selector box around the navigator plot, I also got a bunch of repeated warnings about recursive repaints:
```
QWidget::repaint: Recursive repaint detected
QWidget::paintEngine: Should no longer be called
QPainter::begin: Paint device returned engine == 0, type: 1
QPainter::end: Painter not active, aborted
QWidget::repaint: Recursive repaint detected
QWidget::paintEngine: Should no longer be called
QPainter::begin: Paint device returned engine == 0, type: 1
QPainter::end: Painter not active, aborted
QWidget::repaint: Recursive repaint detected
QWidget::paintEngine: Should no longer be called
QPainter::begin: Paint device returned engine == 0, type: 1
QPainter::end: Painter not active, aborted
```

**Comment by jat255** [comment]
Yup, so the issue is at https://github.com/em-MRCZ/python-mrcz/blob/master/mrcz/__init__.py#L7

If you install `mrcz` locally and comment out that line (disabling the `ReliablePy` stuff), then the UI loads as expected.

Interestingly, with the qt5 backend, the issue of #154 is different. In this case, the window opens, but Windows won't let you save the file because it has a colon and a newline. Better than before, but still not ideal.

**Comment by ericpre** [comment]
I have pushed a few more commits, which should fix the issue with setting the backend by delaying the import of hyperspy because the backend is set. This should work as a workaround but indeed it would be better to fix the issue in hyperspy or mrcz, where there is indeed a `matplotlib.pyplot` import.
Can you try to install hyperspyui using this branch?
Which version of matplotlib do you have?

**Comment by jat255** [comment]
Sure, I'll give it a try.

**Comment by jat255** [comment]
@ericpre I got the following error using your branch:

```
C:\Users\jat
[hyperspyUI_test] λ python -m hyperspyui
Traceback (most recent call last):
  File "C:\anaconda3\envs\hyperspyUI_test\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "C:\anaconda3\envs\hyperspyUI_test\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "c:\users\jat\git_repos\hyperspyui\hyperspyui\__main__.py", line 133, in <module>
    main()
  File "c:\users\jat\git_repos\hyperspyui\hyperspyui\__main__.py", line 112, in main
    from hyperspyui.mainwindow import MainWindow
  File "c:\users\jat\git_repos\hyperspyui\hyperspyui\mainwindow.py", line 56, in <module>
    from hyperspyui.mainwindowhyperspy import MainWindowHyperspy, tr
  File "c:\users\jat\git_repos\hyperspyui\hyperspyui\mainwindowhyperspy.py", line 32, in <module>
    from .mainwindowutillayer import MainWindowActionRecorder, tr
  File "c:\users\jat\git_repos\hyperspyui\hyperspyui\mainwindowutillayer.py", line 37, in <module>
    from hyperspyui import hooktraitsui
  File "c:\users\jat\git_repos\hyperspyui\hyperspyui\hooktraitsui.py", line 25, in <module>
    import traitsui.qt4.ui_base as ui_base
  File "C:\anaconda3\envs\hyperspyUI_test\lib\site-packages\traitsui\qt4\__init__.py", line 18, in <module>
    import pyface.qt
  File "C:\anaconda3\envs\hyperspyUI_test\lib\site-packages\pyface\qt\__init__.py", line 44, in <module>

    % qt_api)
RuntimeError: Invalid Qt API 'pyqt5', valid values are: 'pyqt' or 'pyside'
```

And here are my package lists: https://gist.github.com/jat255/20d76000cd0b2fb8bad2c45ec87cdd3e

**Comment by ericpre** [comment]
This branch requires pyface and traitsui development branches and they are installed in the bundle. If you can't build the pyface and traitsui on your computer, you can try to uninstall and re-install hyperspyui using this branch in the hyperspy bundle. Doing it in the bundle will also allow you not to mess your conda environment! ;)

**Comment by jat255** [comment]
Ah, that explains it. This is a testing environment anyway, so no worries there (the joys of Anaconda!)

I'll get those installed and see how it goes.

**Comment by ericpre** [comment]
em-MRCZ/python-mrcz#8

**Comment by jat255** [comment]
Ok, it loaded this time, but my signal plotter isn't working:

image

There's also no red square on my navigator plot...

Only relevant output in the console is:
```
QWidget::repaint: Recursive repaint detected
QWidget::paintEngine: Should no longer be called
QPainter::begin: Paint device returned engine == 0, type: 1
QPainter::end: Painter not active, aborted
```

**Comment by jat255** [comment]
Relevant versions:

```
hyperspy (1.4.dev0+git.554.g3ea779ae)
hyperspy-gui-traitsui (1.0.3, c:\users\jat\git_repos\hyperspy_gui_traitsui)
hyperspyUI (1.1.0+dev, c:\users\jat\git_repos\hyperspyui)
matplotlib (2.2.2)
pyface (5.2.0.dev347)
QtPy (1.4.0)
traitsui (5.2.0.dev420)

pyqt                      5.6.0                    py35_4    conda-forge
qt                        5.6.2                    vc14_1  [vc14]  conda-forge
```

**Comment by jat255** [comment]
This the result using your branch in the bundle: 

image

Interestingly, the red square does get plotted, and if I move it at all, the signal plot gets drawn; the signal is only blank when the signal first opens.

image

No `matplotlib` backend error though, so improvement!

**Comment by ericpre** [comment]
Yes, this is the issue mentioned at the beginning of this PR. Normally, it should work fine with matplotlib 2.0.

**Comment by jat255** [comment]
Yup, you're right. Downgrading to 2.0.2 works fine for me. Do you know if this is something big that needs to be changed in HyperSpy? Or is it something we can workaround?

**Comment by robbmcleod** [comment]
If you bump your requirement to `mrcz>=0.3.6` you should be good now.

**Comment by francisco-dlp** [comment]
I get the following error in Linux using the latest mpl:

```python
Traceback (most recent call last):
  File "/home/francisco/anaconda/anaconda/bin/hyperspyui", line 11, in <module>
    load_entry_point('hyperspyUI', 'gui_scripts', 'hyperspyui')()
  File "/home/francisco/Git/hyperspyUI/hyperspyui/__main__.py", line 116, in main
    from hyperspyui.mainwindow import MainWindow
  File "/home/francisco/Git/hyperspyUI/hyperspyui/mainwindow.py", line 35, in <module>
    from hyperspyui.mainwindowhyperspy import MainWindowHyperspy, tr
  File "/home/francisco/Git/hyperspyUI/hyperspyui/mainwindowhyperspy.py", line 41, in <module>
    from .mainwindowutillayer import MainWindowActionRecorder, tr
  File "/home/francisco/Git/hyperspyUI/hyperspyui/mainwindowutillayer.py", line 24, in <module>
    from .mainwindowbase import MainWindowBase, tr
  File "/home/francisco/Git/hyperspyUI/hyperspyui/mainwindowbase.py", line 39, in <module>
    import hyperspyui.mdi_mpl_backend
  File "/home/francisco/Git/hyperspyUI/hyperspyui/mdi_mpl_backend.py", line 45, in <module>
    from matplotlib.backends.backend_qt5 import (SPECIAL_KEYS, SUPER, ALT, CTRL,
ImportError: cannot import name 'fn_name'
```

**Comment by ericpre** [comment]
It looks to me, you are on the wrong branch or something similar, because this import is actually removed in this PR!

**Comment by ericpre** [comment]
I have edited the first post to clarify the situation regarding the plotting issue -which is now fixed in https://github.com/hyperspy/hyperspy/pull/1890. :smile:

File: hyperspyui/mainwindow.py [hunk 1]
```diff
@@ -31,9 +31,29 @@
...
+from qtpy.QtWidgets import QApplication, QSplashScreen
+from qtpy.QtGui import QColor, QPixmap
+from qtpy.QtCore import Qt
+splash_pix = QPixmap(os.path.join(
+    os.path.dirname(__file__), 'images', 'splash.png'))
+SPLASH = QSplashScreen(splash_pix, Qt.WindowStaysOnTopHint)
+SPLASH.show()
+SPLASH.showMessage("Initializing...", Qt.AlignBottom | Qt.AlignCenter |
+                   Qt.AlignAbsolute, QColor(Qt.white))
+QApplication.processEvents()
```
Comment (vidartf) [inline hyperspyui/mainwindow.py]: This feels quite inappropriate to have in the global scope (executed on import). Why was this moved here?

File: hyperspyui/mainwindow.py [hunk 2]
```diff
@@ -60,31 +73,24 @@ class MainWindow(MainWindowHyperspy):
...
+        self.splash = SPLASH
+
         # State variables
         self.signal_type_ag = None
         self.signal_datatype_ag = None
         self._plugin_manager_widget = None
 
+        self._load_signal_types()
+
+        print('here')
```
Comment (vidartf) [inline hyperspyui/mainwindow.py]: Debug code

File: hyperspyui/mainwindowbase.py [hunk 3]
```diff
@@ -376,6 +371,17 @@ def select_tool(self, tool):
         self.active_tool = tool
         tool.connect_windows(self.figures)
 
+    def setUpdatesEnabled(self, value, force_setting=False):
+        """
+        pyqt5 doesn't seem to suffer from screen flickering and it is not
```
Comment (vidartf) [inline hyperspyui/mainwindowbase.py]: This comment is not very helpful to those who don't know what type of flickering is mentioned. Is this really needed? What problem does it solve? Also, this text is probably better as a comment than a docstring?

File: hyperspyui/uiprogressbar.py [hunk 4]
```diff
@@ -28,25 +28,29 @@
...
-# This is necessary as it bugs out if not (it's a daisy chained event)
+class Signaler(QObject):
+    created = Signal([int, int, str])
+    progress = Signal([int, int, str], [int, int])
+    finished = Signal(int)
+
+    cancel = Signal(int)
+    # This is necessary as it bugs out if not (it's a daisy chained event)
 
+    def _on_cancel(pid):
```
Comment (vidartf) [inline hyperspyui/uiprogressbar.py]: Won't the first argument here be `self`? Should this be a class/static method?

File: setup.py [hunk 5]
```diff
@@ -37,23 +37,25 @@
     license='GPLv3',
     packages=find_packages(exclude=['tests*',
                                     'hyperspyui.plugins.user_plugins']),
-    requires=['hyperspy (>= 1.1.1)',
+    requires=['hyperspy (>= 1.3.0)',
```
Comment (vidartf) [inline setup.py]: Is 1.3.0 actually required? The only change to hyperspy fronting code I see is in ` hyperspyui/overrides` and that should be possible to put in a `try`/`except: pass` block.

File: setup.py [hunk 6]
```diff
@@ -37,23 +37,25 @@
...
               'autopep8',
+              'pyface', #  (>=6.0.0) for pyqt5
               'traits',
-              'traitsui',
+              'traitsui',  # (>=5.2.0) for pyqt5
               'qtconsole',
+              'qtpy',
               ],
-    install_requires=['hyperspy >= 1.1.1',
+    install_requires=['hyperspy >= 1.3.0',
```
Comment (vidartf) [inline setup.py]: Same here.

File: hyperspyui/mainwindow.py (hunk 1 above)
Comment (vidartf) [reply hyperspyui/mainwindow.py]: Maybe put it in a function that get's called in an appropriate spot?

File: setup.py (hunk 5 above)
Comment (vidartf) [reply setup.py]: Would you mind reverting this to 1.1.1 (here and below)? If we are unsure, it is better to leave it up to the user.

Comment (ericpre) [reply setup.py]: Done!

**Comment by vidartf** [comment]
I'll take it for a spin later and iterate if I find anything else. Thanks for the work!

**Comment by ericpre** [comment]
:+1: 
It looks to me that pyside support is broken, see
https://github.com/hyperspy/hyperspyUI/blob/master/hyperspyui/advancedaction.py#L36

But this is not because of this PR :wink: I have just checked that it was already the case before.

**Comment by jat255** [comment]
Now that installing the latest master makes using qt5 is not really optional, should we include updated requirements in the `setup.py`? I see that there are comments, and a note in the readme, but it's not extremely clear what commands would be necessary to get a working installation.

I would suggest adding the following to the README:
```
pip install git+https://github.com/enthought/pyface@release/6.0.0
pip install git+https://github.com/enthought/traitsui
```
